packages/site/cms-site/src/blockRegistry/blockRegistry.ts

```diff
@@ -0,0 +1,136 @@
+import type { GraphQLClient } from "graphql-request";
+
+type BlockMetaField = {
+    name: string;
+    kind: string;
+    nullable: boolean;
+    enum?: string[];
+    block?: string;
+    blocks?: Record<string, string>;
+    object?: { fields: BlockMetaField[] };
+};
+
+interface BlockMeta {
+    name: string;
+    fields: BlockMetaField[];
+    inputFields: BlockMetaField[];
+}
+
+// we have two typings for block meta
+// 1. for the public api (BlockMeta, BlockMetaField): simpler and compatible with import from block-meta.json (where eg. kind is a string)
+// 2. for the internal api (BetterBlockMeta, BetterBlockMetaField): more complex with more type safety
+type BetterBlockMetaField =
+    | {
+          name: string;
+          kind: "String" | "Number" | "Boolean" | "Json";
+          nullable: boolean;
+      }
+    | {
+          name: string;
+          kind: "Enum";
+          nullable: boolean;
+          enum: string[];
+      }
+    | {
+          name: string;
+          kind: "Block";
+          nullable: boolean;
+          block: string;
+      }
+    | {
+          name: string;
+          kind: "OneOfBlocks";
+          nullable: boolean;
+          blocks: Record<string, string>;
+      }
+    | {
+          name: string;
+          kind: "NestedObject" | "NestedObjectList";
+          nullable: boolean;
+          object: BetterBlockMetaNestedObject;
+      };
+
+interface BetterBlockMeta {
+    name: string;
+    fields: BetterBlockMetaField[];
+    inputFields: BetterBlockMetaField[];
+}
+
+interface BetterBlockMetaNestedObject {
+    fields: BetterBlockMetaField[];
+}
+
+/* eslint-disable @typescript-eslint/no-explicit-any */
+export interface RegisterBlockOptions<BlockData> {
+    loader: (options: { blockData: BlockData; client: GraphQLClient }) => Promise<any> | any;
+}
+const blocks: Record<string, RegisterBlockOptions<any>> = {};
+export function registerBlock<BlockData = unknown>(blockName: string, options: RegisterBlockOptions<BlockData>) {
+    blocks[blockName] = options;
+}
+
+export async function recursivelyLoadBlockData({
+    blockType,
+    blockData,
+    client,
+    blocksMeta,
+}: {
+    blockType: string;
+    blockData: unknown;
+    client: GraphQLClient;
```

Comment:
The client is only passed around here. How about using a generic for this function instead? This would remove the dependency on grapqhl-request in `@comet/cms-site` and would enable using other GraphQL clients (Apollo, urql)?

Reply:
Then you also have a a generic in the block loader function that you always have to specify. Maybe we could use module augmentation?

Reply:
> Maybe we could use module augmentation?

Thought about this too. Maybe even better!